crates/flipperzero/src/gpio/i2c.rs

```diff
@@ -309,13 +309,410 @@ impl BusHandle {
+        .ok_or(Error::TransferFailed)
+    }
+
+    // This function is generic to allow the implementation of both versions of the embedded_hal
+    // transaction traits
+    fn transaction_impl<'a, O>(
+        &mut self,
+        device: DeviceAddress,
+        operations: &mut [O],
+        timeout: Duration,
+    ) -> Result<(), Error>
+    where
+        O: OperationLike + 'a,
+    {
+        use sys::{
+            FuriHalI2cBegin_FuriHalI2cBeginRestart as BeginRestart,
+            FuriHalI2cBegin_FuriHalI2cBeginResume as BeginResume,
+            FuriHalI2cBegin_FuriHalI2cBeginStart as BeginStart,
+            FuriHalI2cEnd_FuriHalI2cEndAwaitRestart as EndAwaitRestart,
+            FuriHalI2cEnd_FuriHalI2cEndPause as EndPause,
+            FuriHalI2cEnd_FuriHalI2cEndStop as EndStop,
+        };
+
+        let mut operations = operations.iter_mut().peekable();
+        let mut start = BeginStart;
+        let address = device.0.into();
+
+        while let Some(op) = operations.next() {
+            let (end, next_start) = match (op.kind(), operations.peek().map(|next| next.kind())) {
+                (OperationKind::Read, Some(OperationKind::Read))
+                | (OperationKind::Write, Some(OperationKind::Write)) => (EndPause, BeginResume),
+                (_, Some(_)) => (EndAwaitRestart, BeginRestart),
+                (_, None) => (EndStop, BeginStart),
+            };
+
+            let result = unsafe {
+                match op.as_op() {
+                    Operation::Read(buffer) => flipperzero_sys::furi_hal_i2c_rx_ext(
+                        self.handle,
+                        address,
+                        false,
+                        buffer.as_mut_ptr(),
+                        buffer.len(),
+                        start,
+                        end,
+                        timeout.as_millis() as u32,
+                    ),
+                    Operation::Write(buffer) => flipperzero_sys::furi_hal_i2c_tx_ext(
+                        self.handle,
+                        address,
+                        false,
+                        buffer.as_ptr(),
+                        buffer.len(),
+                        start,
+                        end,
+                        timeout.as_millis() as u32,
+                    ),
+                }
+            };
+
+            if !result {
+                return Err(Error::TransferFailed);
+            }
+
+            start = next_start;
+        }
+
+        Ok(())
+    }
 }
 
 #[derive(Debug, PartialEq)]
 pub enum Error {
     TransferFailed,
 }
 
+/// I2C operation.
+///
+/// Several operations can be combined as part of a transaction.
+pub enum Operation<'a> {
+    /// Read data into the provided buffer
+    Read(&'a mut [u8]),
+    /// Write data from the provided buffer
+    Write(&'a [u8]),
+}
+
+// These exist to allow compatibility with both embedded_hal 1.0 and 0.2 versions of the Operation
+// enum
+
+enum OperationKind {
```

Comment:
Does it makes sense for these enums to derive the standard `Debug`, `Copy`, `Clone`, `Eq`, `PartialEq`, `Hashable` traits?

Reply:
Because `OperationKind` is only meant for internal use, I don't think it needs them.
I went ahead and derived `Debug`, `Eq` and `PartialEq` for `Operation`, mirroring the version found in [embedded-hal](https://docs.rs/embedded-hal/1.0.0-rc.1/src/embedded_hal/i2c.rs.html#279).

Reply:
Sounds reasonable to me.